Gems/MachineLearning/Code/Source/Models/MultilayerPerceptron.h

```diff
@@ -0,0 +1,111 @@
+    class MultilayerPerceptron
+        : public INeuralNetwork
+    {
+    public:
+
+        AZ_RTTI(MultilayerPerceptron, "{E12EF761-41A5-48C3-BF55-7179B280D45F}", INeuralNetwork);
+
+        //! AzCore Reflection.
+        //! @param context reflection context
+        static void Reflect(AZ::ReflectContext* context);
+
+        MultilayerPerceptron();
+        MultilayerPerceptron(const MultilayerPerceptron&);
+        MultilayerPerceptron(AZStd::size_t activationCount);
+        virtual ~MultilayerPerceptron();
+
+        MultilayerPerceptron& operator=(const MultilayerPerceptron&);
+
+        //! INeuralNetwork interface
+        //! @{
+        AZStd::string GetName() const override;
+        AZStd::string GetAssetFile(AssetTypes assetType) const override;
+        AZStd::size_t GetInputDimensionality() const override;
+        AZStd::size_t GetOutputDimensionality() const override;
+        AZStd::size_t GetLayerCount() const override;
+        AZ::MatrixMxN GetLayerWeights(AZStd::size_t layerIndex) const override;
+        AZ::VectorN GetLayerBiases(AZStd::size_t layerIndex) const override;
+        AZStd::size_t GetParameterCount() const override;
+        IInferenceContextPtr CreateInferenceContext() override;
+        ITrainingContextPtr CreateTrainingContext() override;
+        const AZ::VectorN* Forward(IInferenceContextPtr context, const AZ::VectorN& activations) override;
+        void Reverse(ITrainingContextPtr context, LossFunctions lossFunction, const AZ::VectorN& activations, const AZ::VectorN& expected) override;
+        void GradientDescent(ITrainingContextPtr context, float learningRate) override;
+        bool LoadModel() override;
+        bool SaveModel() override;
+        //! @}
+
+        //! Adds a new layer to the model.
+        void AddLayer(AZStd::size_t layerDimensionality, ActivationFunctions activationFunction = ActivationFunctions::ReLU);
+
+        //! Retrieves a specific layer from the model, this is not thread safe and should only be used during unit testing to validate model parameters.
+        Layer* GetLayer(AZStd::size_t layerIndex);
+
+        //! Base serialize method for all serializable structures or classes to implement.
+        //! @param serializer ISerializer instance to use for serialization
+        //! @return boolean true for success, false for serialization failure
+        bool Serialize(AzNetworking::ISerializer& serializer);
+
+        //! Returns the estimated size required to serialize this model.
+        AZStd::size_t EstimateSerializeSize() const;
+
+    private:
+
+        void OnActivationCountChanged();
+
+        //! The model name.
+        AZStd::string m_name;
+
+        //! The model asset file.
+        AZStd::string m_modelFile;
+
+        //! Optional test and train asset data files.
+        AZStd::string m_testDataFile;
+        AZStd::string m_testLabelFile;
+        AZStd::string m_trainDataFile;
+        AZStd::string m_trainLabelFile;
```

Comment:
I'm really not a fan of the way this circumvents the asset system. It prevents asynchronous loads, hot-reloading, auto-detecting of the asset dependencies for inclusion in seed lists, etc. What's the reasoning behind not making a simple asset handler around the LoadModel code?

Also, because these paths are strings, am I correct in thinking they don't show file picker dialogs via the Edit Context?

Reply:
Yup, they're just raw strings.. One thing this allowed that turned out to be really useful was I could save a model, try a refinement training session and just observe the models performance on test and train data, and if I didn't get the result expected I could instantly restore and try a different training approach. I'd like to keep that flexibility as it was *super* useful.